### ISSUE TITLE
Allow editing of posts

### ISSUE THREAD

**Issue by madrobby** [bluesky-social/social-app#673]
**Is your feature request related to a problem? Please describe.**

I make typos and everybody does, and we should be able to fix it. We also make factual or content errors and should be able to correct ourselves; as well as update posts that reflect safety or other time-critical issues. (See below for more safety/accessibility notes.)

**Describe the solution you'd like**

Editing of posts like on Twitter (paid version) or on recent Mastodon versions.
Edited post should be clearly marked with a history of edits available to view.

**Describe alternatives you've considered**

Delete & Repost (like on Mastodon), but that loses replies/embeds/connections/quotes, etc. However, this could be done in the client as a nice first step as it doesn't need backend/storage/schema changes.

**Additional context**
- This is a safety issue: certain posts will benefits from updates (e.g. "Update: The shooter has been arrested.") when people see reposts later.
- This is an accessibility issue: people who have mental or physical limitations with typing are more prone to errors.

**Comment by pfrazee** [comment]
This is on the roadmap, but for us to implement the edit history we need to update the protocol to include a new kind of backlink to previous versions of a record. We'll prioritize this after other core parts of the protocol stabilize.

**Comment by mschwendener** [comment]
fwiw: in some forums, you get a certain amount of minutes to edit. I've recently seen a countdown from 60 minutes. 

But of course that would not cover all cases.

**Comment by PtrJsn** [comment]
> This is on the roadmap, but for us to implement the edit history we need to update the protocol to include a new kind of backlink to previous versions of a record. We'll prioritize this after other core parts of the protocol stabilize.

If the edit history is the only thing blocked, I feel it would be valuable to implement editing just with an indicator that it's been edited, & come back to the history display later.

**Comment by Sininini** [comment]
I would like to see editing of ALT text and outside tags first. I don't think these would really need history either so should be a bit easier to implement.

**Comment by venteto** [comment]
Once posts are editable and have a public edit history available, a cool bonus enhancement would be if the original poster could see all of their post interactions (such as views, likes, reposts, quote posts, or any other future analytics metrics) attached to / segmented by the specific post ~version~ that was viewed by the account generating the interaction.

Then for example you could see exactly who all saw your post _before_ you caught your ten typos and fixed them, or before you added the link or hashtag you forgot (and potentially had to cannibalize actual text from your post in order to add retroactively, while these things still consume characters from actual post text).

**Comment by venteto** [comment]
One thing I absolutely think is a terrible idea is setting any arbitrary time limits for editing posts. When you have a visible post edit history, such limits are not even necessary. Moreover, there is a burden on the user making the post, in wanting to be able to modify it as necessary without losing things attached to it such as likes, replies, reposts, et al., but there's no similar burden on any other user reading the post. As long as other users can see the edit history, it's really not reaasonable for them to impose (or ask the devs to impose) a limit of n seconds / minutes / hours / days within which the post can be edited.

Notably, on Twitter when I would click a hashtag, I fully expected to be able to see posts that were years old among other more recent posts (such is the the value of a hashtag for kind related content, regardless of age). So on Bluesky, if 6 or 12 months from now I think of another hashtag that would be appropriate for an OLD post, I'd like to be added (but this specific genre of edit also touches upon a separate issue — namely, hashtags & mentions should be in, for lack of a better description, separate database tables with a many to many relation, not directly in the content of the post itself ... these thinfs are _not_ content at all ... but now I'm going on a tangent ...)

Another good feauture to add with the addition of editable posts and viewable post edit histories would be an in-app notification system, whereby any user who previously had interacted with a given post would just receive a notification when that post is edited, and then they can decide if te edit really affects their previous interaction or not, and respond accordingly if they feel like it does. This especially would be good for journalists who have accounts, and had previously also embedded a post (when this becomes a feature too) in an offsite news article. Then a journalist or editor could decide if they wanted to retain the embedded post in their article, add a bit of "update" text to their article, etc.

Incidentally, everything pertaining to a post should be editable, including the alt text on attached images.

**Comment by Lucent** [comment]
Would, or could, editing of posts include retroactively turning them into a thread--a reply to another of our own older posts, without textual editing?

**Comment by venteto** [comment]
@Lucent interesting idea ... makes me think of the simplicity of changing the _parent_ of a "label" in Gmail ... could be useful in the microblog paradigm that makes "threads" necessary at all

**Comment by okpierre** [comment]
I can see the pros/cons of this. I'll vote for post edit with history and includes time limit
-with time limit = less notifications
-with time limit = less history retention needed

**Comment by snarfed** [comment]
Federation may make this issue more acute, at least for rendering edits if not providing a UI to make them. PDSes can happily allow editing, mine does, but bsky.app doesn't seem to show edits/updates. It always only shows the original record.

Asn an example, I created and then edited a post on my PDS. [atproto.tools shows both the create and the update as valid](https://atproto.tools/records?did=did%3Aplc%3Abg5udl25mvzg3rt7l5n2hzet&collection=app.bsky.feed.post&rkey=3krh65hkwbp42), but [bsky.app only shows the original create](https://bsky.app/profile/snarfed.indieweb.social.ap.brid.gy/post/3krh65hkwbp42).

**Comment by enn-nafnlaus** [comment]
> I can see the pros/cons of this. I'll vote for post edit with history and includes time limit -with time limit = less notifications -with time limit = less history retention needed

If I may,  I think that people are...

~ ｏνＥ𝔯𝐂𝕆๓𝓟Ļι𝔠ａＴι𝓷Ǥ    丅𝔥𝕖    𝐢ᔕѕᑌⓔ  ~

.... ;)  

Bluesky's principle has always been "_let users decide_", right?  So, fine.  Keep an edit history, and let the *user decide* which version they want to see, whether that's the version that was posted with 0 minutes, 5 minutes, 60 minutes, infinite minutes.

Now everyone is happy, right?

And re: sharing posts, isn't the solution obviously just "the preview that shows up of the shared post is the version that was present at the time they clicked 'share' - and then if they click through, they see the most recent one?  I mean, we could *also* let users configure what preview they want to see on shared posts, but at a most basic level, wouldn't that be good enough for most people?

**Comment by zaphbbrox** [comment]
> Federation may make this issue more acute, at least for rendering edits if not providing a UI to make them. PDSes can happily allow editing, mine does, but bsky.app doesn't seem to show edits/updates. It always only shows the original record.
> 
> Asn an example, I created and then edited a post on my PDS. [atproto.tools shows both the create and the update as valid](https://atproto.tools/records?did=did%3Aplc%3Abg5udl25mvzg3rt7l5n2hzet&collection=app.bsky.feed.post&rkey=3krh65hkwbp42), but [bsky.app only shows the original create](https://bsky.app/profile/snarfed.indieweb.social.ap.brid.gy/post/3krh65hkwbp42).
> 
> cc snarfed/bridgy-fed#947

Strong agreement from my side. I just had to delete and rewrite my federated mastodon post due to the lower char limit on Bluesky and the missing edit option for three times.
Would be nice to have at least a temporary solution (e.g. without saving the history of an edited post).

**Comment by jeffmcneill** [comment]
How was editing not in the alpha versions? How has it be "de-prioritized"? Mastodon has had this for 2+ years https://blog.joinmastodon.org/2022/03/mastodon-3.5/

**Comment by surfdude29** [comment]
> How was editing not in the alpha versions? How has it be "de-prioritized"? Mastodon has had this for 2+ years https://blog.joinmastodon.org/2022/03/mastodon-3.5/

Twitter added the ability to edit posts (for some users) in 2022, 16 years after it launched.

Mastodon also added the ability to edit posts in 2022, 5 years after it launched.

Bluesky, meanwhile, only opened up to the public in 2023, so it hasn't even been 2 years. We'll get there, but the devs can't add every feature requested all at the same time.

**Comment by Mattis142** [comment]
> > How was editing not in the alpha versions? How has it be "de-prioritized"? Mastodon has had this for 2+ years https://blog.joinmastodon.org/2022/03/mastodon-3.5/
> 
> Twitter added the ability to edit posts (for some users) in 2022, 16 years after it launched.
> 
> Mastodon also added the ability to edit posts in 2022, 5 years after it launched.
> 
> Bluesky, meanwhile, only opened up to the public in 2023, so it hasn't even been 2 years. We'll get there, but the devs can't add every feature requested all at the same time.

threads shipped editing 3 months after launching

**Comment by erlend-sh** [comment]
> threads shipped editing 3 months after launching

On the other hand they *still* don’t have proper two-way integration with other ActivityPub instances. These apps are working with very different constraints and priorities.

**Comment by surfdude29** [comment]
> > threads shipped editing 3 months after launching
> 
> On the other hand they _still_ don’t have proper two-way integration with other ActivityPub instances. These apps are working with very different constraints and priorities.

Indeed. Also, Meta said that Threads had a team of [about 60 engineers](https://engineering.fb.com/2023/09/07/culture/threads-inside-story-metas-newest-social-app/) and it's almost certainly grown in the year since that article was published. Afaik Bluesky currently only has 12 engineers working on it.

**Comment by enn-nafnlaus** [comment]
What was done in the past is not a productive line of discussion.  Getting this done in the future is.  :)  We know what's not been done.  Let's focus on getting it done.

**Comment by surfdude29** [comment]
> What was done in the past is not a productive line of discussion. Getting this done in the future is. :) We know what's not been done. Let's focus on getting it done.

That’s a good point. For those who haven’t seen it, it might be helpful to quote from the [2024 Protocol Roadmap](https://docs.bsky.app/blog/2024-protocol-roadmap#on-the-horizon) (published May 6, 2024), which as far as I know is the latest “official” thinking from Bluesky on their plans for post editing:

> **On the Horizon**
>
> There are a few other pieces of protocol work which we are starting to plan out, but which are not currently scheduled to complete in 2024. It is very possible that priorities and schedules will be shuffled, but we mostly want to call these out as things we do want to complete, but will take a bit more time.
>
>…
>
> **Record Versioning (Post Editing):** atproto already supports updating records in repositories: one example is updating bsky profile records. And preparations were made early in the protocol design to support post editing while avoiding misleading edits. Ideally, it would also be possible to (optionally) keep old versions of records around in the repository, and allow referencing and accessing multiple versions of the same record.

**Comment by enn-nafnlaus** [comment]
The above examples aren't just theoretical, and to be fair, don't even require editing. It was admittedly quite amusing on Twitter, for example, when Dmitri Rogozin, a Russian Nazi and member of Putin's government, quote-tweeted someone who criticized him with text like "Based on your profile photo, you're a f***ot", and the person responded by changing their profile picture to a picture of Rogozin  ;)  But this is ideally the sort of thing isn't something we don't want happening.  Linked records of all types really need to be timestamped.

**Comment by surfdude29** [comment]
> Do you know if there's any way to leave comments on it?

You can leave comments on the roadmap in [the linked GitHub conversation](https://github.com/bluesky-social/atproto/discussions/2469) in the atproto discussion forum.

In addition to or instead of that, you might want to create a new conversation thread specifically about post editing. I imagine the [Protocol (atproto) section](https://github.com/bluesky-social/atproto/discussions/categories/protocol-atproto ) might be best, but that’s up to you :)

**Comment by verdverm** [comment]
fwiw, you can edit records already, and while Bluesky doesn't make it available in the App View or display edits, you can make this work with custom lexicon or alternative UIs today. There are a lot of UI affordances and social contract stuff to think about, but the beauty of ATProtocol is that we can have multiple, competing takes on what those should be.

I would advocate for protocol level changes that make the scheme I've outline into atomic edits, because of race conditions

https://verdverm.com/blog/adding-record-editing-with-history-to-atprotocol

**Comment by Turnitup5000db** [comment]
So I'm not a programmer but I posted a conceptual suggestion to the app that someone there suggested I post here.

My thought was that it might be possible to avoid stealth edits and still allow editing by, when an edit occurs, replacing the original post with the edited version, and demoting the original post and all responses to being a response to the new post, maybe with some kind of tagging system as well.

I hope that all makes sense and is useful.  As I said, this isn't my area of expertise and I have no idea of the technical obstacles or problems involved.

**Comment by verdverm** [comment]
@Turnitup5000db the technical side is trivial compared to the social complexities. Some of them are sketched out in prior comments, however there are a lot of adversarial situation to think through as well as some that I would not consider adversarial

1. What happens if reply to a post saying I support the idea and the author changes it so something I don't support?
1. What happens to the like count? (keep or reset)
1. Should some of these decisions be left up to the user, given how user choice centric ATProto is in ethos?

**Comment by venteto** [comment]
> [@Turnitup5000db](https://github.com/Turnitup5000db) the technical side is trivial compared to the social complexities. Some of them are sketched out in prior comments, however there are a lot of adversarial situation to think through as well as some that I would not consider adversarial (like fixing a typo or mispelling, how would you know if an edit crosses some threshold?)

I think all metadata attached to posts AND replies should be split up into A) cumulative and B) historical batches. So if you make a post, 10 people like it, then you edit it, and 5 more people like it therafter, then you edit it again, and another 5 people like it thereafter.

 then metadata should show you have 5 current likes, 20 cumulative likes, and 15 historical likes that are grouped according to the post / reply revision they were originally attached too. Additionally, if one of your revisions causes previous likers to remove their likes, and additional bit of metadata for removed likes could also be shown, indicating that the revision might have been controversial (but future readers can decide for themselves whether the edit was controversial, since no one can know for certain why the likes were rescinded).

One fundamental feature I think should go hand in hand with editing posts (and replies) is for any user who in any way has interacted with a post (or reply) to receive a notification anytime an update is made to anything they have interacted with (e.g., liked, reposted, quote-posted, replied to, hidden, reported, muted, bookmarked).

So if you get a notification, then you can decide if you want to alter any of your previous interactions or actions with the post, reply, or user. And along with the notification should come a requirement to click sometihing to acknowlege you want your like, comment, et al., to apply to the current version of the post / reply, so that your place in the post / reply metadata can be moved up to the metadata for the *current* post / reply version, otherwise if the interacting user ignores this, then their previous interactions are left in the metadata counts for the respective version they interacted with.

In post editing windows You could also possibly add a check box like Wikimedia saying "this is a minor edit" (e.g., for just fixing typos). Then there could perhaps be an item in our account settingss to say "automatically advance my post / reply interactions to the current revision if an edit is a minor edit". This is in case it is too bothersome for users to go manually acknowledge that an edit has taken place, and it can certainly be abused, but ultimately it is stll up to the interacting user to look at current versions of posts and replies and make sure they want their interactions to still apply. So if you liked a post which said "I love orchids", then the user who posted it makes an edit to say "I HATE orchids!" and dishonestly checks the "minor edit" box, users as a whole could still also have the communal opportunity to flag the post / reply for "edit abuse" even if a given interactor  has A) configured their account settings to auto-advance their content interactions to current content revisions and B) also ignores the fact that the post / reply has not ontly been editing, but its meaning fundamentally alterted. You have the opportunity to remove your like, or prevent your user settings from advancing your like to the current version, but no one can make you take either course of action, and other users can still moderate other users who routinely abuse any "minor edit" functionality.

All of these changes should be protocol level. Other clients should be compliant with protocol-level features. If your alternative client is hiding these features from users, your client should be banned accessing data on this network, and should also be added to a blacklist of undesirable clients if you decide to also just implement your own network for your client to consume and lure unsuspecting new users to. There's no beneficial reason to balkanize this stuff, and any protocol is inherently opinionated, and these features benefit users, they are not mere a la carte garnishes for alt client / alt network devs to ignore, because why would you want to not benefit your users?

**Comment by verdverm** [comment]
> All of these changes should be protocol level.

I don't agree with this. While it may make sense for Bluesky records, we can build anything on ATProto and it will not make sense for all applications. The post, likes, and the graph are not even protocol level, they are records of Bluesky lexicon a layer above the protocol. Not all record types have user like and reply interactions.

> your client should be banned accessing data on this network

This is kind of against the open ethos of ATProto and removes the competition in features and affordances. This is because what you are talking about are semantics a level up from the protocol on specific record types and behaviors. Certainly all clients should comply with the protocol spec. I think Bluesky will make their choice, many Bluesky 3rd party clients will strive to keep parity, while others will not. This is ok in my mind. Let the people decide which set of features and affordances they want rather than forcing one corporations choice on everyone (i.e. this is Bluesky's commitment to a credible exit)

I'm not sure how technically possible it is. How would you know what client is making the API calls to the public endpoints and PDS instances?

**Comment by DHager** [comment]
A partial workaround would be a delayed-send option on clients (app, website) so that a user can choose a delay (e.g. 10 minutes) before it really goes out. This gives users a chance to glance over their pending post and recognize typos or stop accidental early-submits.

**Comment by verdverm** [comment]
3rd party apps and frontends could implement this today on the current spec, though the official client will (1) ignore the edits and (2) will make the copied-on-write records look like new posts. Not something I would really recommend 3rd party Bluesky clients implement as it would break user expectations and experience, but 3rd party appviews could certainly build around this capability

https://verdverm.com/blog/adding-record-editing-with-history-to-atprotocol

**Comment by venteto** [comment]
... and relying on a third-party client for any feature at all inherently means the brag that Bluesky won't use users' content for training AI is still 100% hollow ... because there is a public API allowing just that, all day long, so they might as well retract any post bragging about that

**Comment by venteto** [comment]
Instead of giving an empty thumbs down, why don't you refute what I said and show me how Bluesky posts cannot be used by third-parties to train AI? Can you?

**Comment by surfdude29** [comment]
@venteto The thumbs down might be because your comment was off-topic? This issue is about editing posts, not about the use of Bluesky posts to train AI.

**Comment by mschwendener** [comment]
> A partial workaround would be a delayed-send option on clients (app, website) so that a user can choose a delay (e.g. 10 
> minutes) before it really goes out. This gives users a chance to glance over their pending post and recognize typos or stop 
> accidental early-submits.

What I would need is an '**unsend**' option like I have for email. 

Reason: only after sending posts with links I see how the preview images really look - and can decide if I want to suppress the preview and just keep the link.

**Comment by surfdude29** [comment]
Big news on post editing: Bluesky Head of Product, Alex Benzer, has posted a thread discussing it:

https://bsky.app/profile/alexbenzer.com/post/3mlvrfzti7k25

It's worth reading the whole thread and giving feedback over there, but here are a few highlights:

> we're discussing *potentially* adding post editing to Bluesky this year.

> post editing is easy to build, but it adds a bunch of complexity and risk that's hidden until you look below the surface.

> there are various safety and spam risks, but the core problem is this: someone writes "puppies are the best 🐶", racks up engagement, then edits it to say something horrible or swaps in a link to a malicious site. now people look like they endorsed it.

> this gets especially ugly with reposts and quote posts.
>
> plus, post editing adds more complexity and required tooling for our T&S team.
>
> so if we do this, we'll need guardrails.

> here's what we're thinking:
>
> 1) you have 5 minutes to edit a post
> 2) you can only edit each posts once
> 3) images and videos can’t be edited
> 4) a prominent "edited" badge appears on the post
> 5) version history is visible to everyone
> 6) your account has to be at least a week old

> some of these are tough tradeoffs. 5 minutes won't catch the typo you spot two hours later. one edit means you can't fix the fix. we know. it's the price of mitigating these risks and keeping people safe here. and this is what might be feasible for our small team.

**Comment by mschwendener** [comment]
Thank you for the update!

**Comment by venteto** [comment]
> there are various safety and spam risks, but the core problem is this: someone writes "puppies are the best 🐶", racks up engagement, then edits it to say something horrible

**So then why not apply engagement only to a specific post VERSION**, not to the generic idea of an unversioned post?

So someone (I JUST EDITED THIS COMMENT TO CORRECT MY TYPO FROM 'SOME' TO 'SOMEONE' ... _this works just fine right here on GitHub, this idea is not new_) initally posts "puppies are great", gets 20 likes for that version, and that it what is shown in the feed and on their profile. Then they edit the post to say "puppies are bad" ... this creates a new revision, which ideally would be like a container for the current content of the post, and anything else pertaining to the post, like engagement metrics. From this point forward, until the next revision, all engagement is now associated with this version, and past engagement is essentially archived, and not visible until you click some sort of post edit history widget.

Beyond that, you could also optionally notify everyone who engaged with the earlier revision, so that they have the agency to disavow any approval of the now-revised post content.

**Comment by aaribaud** [comment]
Re: delays: rather than allowing edits *until* some delay has elapsed, I would suggest throttling them:
* by amount of change: small edits would not be throttled much, big edits would be throttled a lot (this could be done by computing an allowed edit date based on date and content of current post and delta between current post and edit, and refusing edit if current date is not past allowed date) ;
* by frequency: series of edits would be increasingly throttled, so that  it would be prohibitive to disguise a big edit (this could be done by adding exponential backoff delay to allowed date based on total count of edits).

This plus a clear marker on posts which have been edited, and public-visible edit history, should cover pretty much every need.

**Comment by pjvjay** [comment]
Looking forward to this as its quite painful when small mistakes are made and you need to manually delete and repost entirely. One suggestion could be a "re-write" feature vs an edit feature. there are so many times when I just needed to re-write the post (delete and repost it with most of the previous content which will lose its likes and reference to parent posts). Its a simple change but it allows for an intermediate fix while the long-term fix and conditions are finalised.

**Comment by pawamoy** [comment]
Is that an issue specific to Bluesky, or an issue for the AT Protocol itself? If the former, apologies, my next sentence is irrelevant, and please let me know where I should comment instead.

Coming from Mastodon / ActivityPub, not being able to edit a post on an atproto view/app (mu.social in my case) feels like terrible UX. On Mastodon I can edit a post, whenever I want, how many times I want. When I do, people who "liked" or quoted the post get a notification that I changed it, so they can review and remove their engagement on it. That's it. Works well. I have no idea how complex it is to implement, but it's good UX to me 🙂
